### PR TITLE
[TACHYON-621] Enable Client PLAIN Authentication

### PR DESCRIPTION
**Pull request by sundapeng**
Jira: https://tachyon.atlassian.net/browse/TACHYON-621

This patch is about Enable Client PLAIN Authentication, including adding Plain Client Transport and callback handler in client side.

Related jira: [TACHYON-726](https://tachyon.atlassian.net/browse/TACHYON-726) Add methods of creating TTransport

### CODE REVIEW

**Comment by gpang** [comment]
ok to test

File: common/src/main/java/tachyon/security/PlainSaslHelper.java [hunk 1]
```diff
@@ -29,10 +40,50 @@
...
+    return new TSaslClientTransport("PLAIN", null, null, null, new HashMap<String, String>(),
+        new PlainCallbackHandler(username, password), wrappedTransport);
+  }
+
+  public static class PlainCallbackHandler implements CallbackHandler {
+
+    private final String mUserName;
+    private final String mPassword;
+
+    public PlainCallbackHandler(String mUserName, String mPassword) {
```
Comment (apc999) [inline common/src/main/java/tachyon/security/PlainSaslHelper.java]: Constructor parameter `mUserName` and `mPassword` is inconsistent with Tachyon naming convention. `userName` and `password` are good. We only make class members like `mUserName`.

File: common/src/main/java/tachyon/security/PlainSaslHelper.java [hunk 2]
```diff
@@ -29,10 +40,50 @@
...
+
+  /**
+   * Get a ClientTransport for Plain
+   * @param username User Name of PlainClient
+   * @param password Password of PlainClient
+   * @param wrappedTransport The original Transport
+   * @return Wrapped Transport with Plain
+   * @throws SaslException
+   */
+  public static TTransport getPlainTransport(String username, String password,
```
Comment (apc999) [inline common/src/main/java/tachyon/security/PlainSaslHelper.java]: Should this be called `getPlainClientTransport` ?

**Comment by apc999** [comment]
I left a few minor comments. Thanks for the patch!
One thing needs to notice is this PR will probably conflict with #1229, let's merge PRs carefully.

File: common/src/main/java/tachyon/security/PlainSaslHelper.java (hunk 1 above)
Comment (sundapeng) [inline common/src/main/java/tachyon/security/PlainSaslHelper.java]: @apc999  Thank you for your comments, I have fix it.

File: common/src/main/java/tachyon/security/PlainSaslHelper.java (hunk 2 above)
Comment (sundapeng) [inline common/src/main/java/tachyon/security/PlainSaslHelper.java]: @apc999  Yes, it should be called *getPlainClientTransport*, thank you for your comments.

**Comment by sundapeng** [comment]
Thank @apc999 for your review, I have updated the patch.

File: common/src/main/java/tachyon/master/MasterClient.java [hunk 3]
```diff
@@ -225,6 +227,28 @@ public synchronized void connect() throws IOException {
   }
 
   /**
+   * Create transport per the connection options Supported transport options are: NOSASL, SIMPLE,
+   * CUSTOM, KERBEROS
+   * @throws TTransportException
+   */
+  private TTransport createTransport() throws IOException {
```
Comment (ooq) [inline common/src/main/java/tachyon/master/MasterClient.java]: Can we put this method in a helper class, maybe `PlainSaslHelper` ? The implementation should be not in this `MasterClient` class.

File: common/src/main/java/tachyon/master/MasterClient.java [hunk 4]
```diff
@@ -225,6 +227,28 @@ public synchronized void connect() throws IOException {
   }
 
   /**
+   * Create transport per the connection options Supported transport options are: NOSASL, SIMPLE,
+   * CUSTOM, KERBEROS
+   * @throws TTransportException
```
Comment (ooq) [inline common/src/main/java/tachyon/master/MasterClient.java]: Please also complete the javadoc with what is returned.

File: common/src/main/java/tachyon/master/MasterClient.java [hunk 5]
```diff
@@ -225,6 +227,28 @@ public synchronized void connect() throws IOException {
   }
 
   /**
+   * Create transport per the connection options Supported transport options are: NOSASL, SIMPLE,
```
Comment (ooq) [inline common/src/main/java/tachyon/master/MasterClient.java]: "." between "options" and "Supported".

File: common/src/main/java/tachyon/master/MasterClient.java [hunk 6]
```diff
@@ -225,6 +227,28 @@ public synchronized void connect() throws IOException {
...
+   * CUSTOM, KERBEROS
+   * @throws TTransportException
+   */
+  private TTransport createTransport() throws IOException {
+    TTransport tTransport = AuthenticationFactory.createTSocket(mMasterAddress);
+    AuthType authType = AuthenticationFactory.getAuthTypeFromConf(mTachyonConf);
+    switch (authType) {
+      case NOSASL:
+        return new TFramedTransport(tTransport);
+      case SIMPLE:
```
Comment (ooq) [inline common/src/main/java/tachyon/master/MasterClient.java]: If we want to fall through here, add a comment for clarification.

File: common/src/main/java/tachyon/master/MasterClient.java (hunk 4 above)
Comment (ooq) [inline common/src/main/java/tachyon/master/MasterClient.java]: Javadoc says throwing `TTransportException` while the method throws `IOException`.

File: common/src/main/java/tachyon/security/PlainSaslHelper.java [hunk 7]
```diff
@@ -29,10 +40,50 @@
...
     return Security.getProvider(PlainSaslServerProvider.PROVIDER) != null;
   }
+
+  /**
+   * Get a ClientTransport for Plain
+   * @param username User Name of PlainClient
+   * @param password Password of PlainClient
+   * @param wrappedTransport The original Transport
+   * @return Wrapped Transport with Plain
+   * @throws SaslException
```
Comment (ooq) [inline common/src/main/java/tachyon/security/PlainSaslHelper.java]: Please explain when the exception will be thrown

File: common/src/main/java/tachyon/security/PlainSaslHelper.java [hunk 8]
```diff
@@ -29,10 +40,50 @@
...
+   */
+  public static TTransport getPlainClientTransport(String username, String password,
```
Comment (ooq) [inline common/src/main/java/tachyon/security/PlainSaslHelper.java]: Note that `PlainSaslHelper` is actually about server side PlainSasl if you read the javadoc and other methods. So I'm not entirely sure if this `getPlainClientTransport` should be here.

File: common/src/main/java/tachyon/security/PlainSaslHelper.java [hunk 9]
```diff
@@ -29,10 +40,50 @@
...
+
+    @Override
+    public void handle(Callback[] callbacks) throws IOException, UnsupportedCallbackException {
+      for (Callback callback : callbacks) {
+        if (callback instanceof NameCallback) {
+          NameCallback nameCallback = (NameCallback) callback;
+          nameCallback.setName(mUserName);
+        } else if (callback instanceof PasswordCallback) {
+          PasswordCallback passCallback = (PasswordCallback) callback;
+          passCallback.setPassword(mPassword.toCharArray());
```
Comment (ooq) [inline common/src/main/java/tachyon/security/PlainSaslHelper.java]: Does it work if mPassword is null?

File: common/src/main/java/tachyon/security/authentication/AuthenticationFactory.java [hunk 10]
```diff
@@ -111,5 +118,27 @@ public static AuthType getAuthTypeFromConf(TachyonConf conf) {
...
+    return new TSocket(NetworkAddressUtils.getFqdnHost(address), address.getPort());
+  }
+
+  /**
+   * Create a ClientTransport for Plain
+   * @param username User Name of PlainClient
+   * @param password Password of PlainClient
+   * @param tTransport The original Transport
+   * @return Wrapped Transport with Plain
+   * @throws SaslException
```
Comment (ooq) [inline common/src/main/java/tachyon/security/authentication/AuthenticationFactory.java]: Please explain when teh SaslException will be thrown.

File: common/src/main/java/tachyon/security/authentication/AuthenticationFactory.java [hunk 11]
```diff
@@ -111,5 +118,27 @@ public static AuthType getAuthTypeFromConf(TachyonConf conf) {
...
+   */
+  public static TTransport createPlainClientTransport(String username, String password,
+      TTransport tTransport) throws SaslException {
+    return PlainSaslHelper.getPlainClientTransport(username, password, tTransport);
```
Comment (ooq) [inline common/src/main/java/tachyon/security/authentication/AuthenticationFactory.java]: Any reason we want this indirection? Can we either, implement `PlainSaslHelper#getPlainClientTransport` in `AuthenticationFactory` or let `PlainSaslHelper#getPlainClientTransport` to be called directly?

**Comment by ooq** [comment]
@sundapeng Thanks for the new PR! I left some initial comments.

File: common/src/main/java/tachyon/master/MasterClient.java (hunk 3 above)
Comment (sundapeng) [inline common/src/main/java/tachyon/master/MasterClient.java]: Fixed

File: common/src/main/java/tachyon/master/MasterClient.java (hunk 4 above)
Comment (sundapeng) [inline common/src/main/java/tachyon/master/MasterClient.java]: Fixed

File: common/src/main/java/tachyon/master/MasterClient.java (hunk 5 above)
Comment (sundapeng) [inline common/src/main/java/tachyon/master/MasterClient.java]: Fixed

File: common/src/main/java/tachyon/master/MasterClient.java (hunk 6 above)
Comment (sundapeng) [inline common/src/main/java/tachyon/master/MasterClient.java]: fixed

File: common/src/main/java/tachyon/security/PlainSaslHelper.java (hunk 7 above)
Comment (sundapeng) [inline common/src/main/java/tachyon/security/PlainSaslHelper.java]: fixed

File: common/src/main/java/tachyon/security/PlainSaslHelper.java (hunk 8 above)
Comment (sundapeng) [inline common/src/main/java/tachyon/security/PlainSaslHelper.java]: getPlainClientTransport could be add here. PlainSaslHelper also provides methods to generate PLAIN transport for server and client. Already added to javadoc

File: common/src/main/java/tachyon/security/PlainSaslHelper.java (hunk 9 above)
Comment (sundapeng) [inline common/src/main/java/tachyon/security/PlainSaslHelper.java]: Good point, fixed

File: common/src/main/java/tachyon/security/authentication/AuthenticationFactory.java (hunk 10 above)
Comment (sundapeng) [inline common/src/main/java/tachyon/security/authentication/AuthenticationFactory.java]: Fixed

File: common/src/main/java/tachyon/security/authentication/AuthenticationFactory.java (hunk 11 above)
Comment (sundapeng) [inline common/src/main/java/tachyon/security/authentication/AuthenticationFactory.java]: Fixed, Move the getClientTransport(InetSocketAddress serverAddress) to AuthenticationFactory. AuthenticationFactory will handle different AuthType.
PlainSaslHelper will handle the Plain related logic.

**Comment by sundapeng** [comment]
Thank @ooq  for your detail review, I have update patch and rebase patch with https://tachyon.atlassian.net/browse/TACHYON-726 Add methods of creating TTransport  and its pr https://github.com/amplab/tachyon/pull/1229

File: common/src/main/java/tachyon/security/authentication/AuthenticationFactory.java [hunk 12]
```diff
@@ -137,6 +143,37 @@ public TTransportFactory getServerTransportFactory() throws SaslException {
...
+   * @param serverAddress the server address which clients will connect to
+   * @return a TTransport for client
+   * @throws IOException if building a TransportFactory fails or user login fails
+   */
+  public TTransport getClientTransport(InetSocketAddress serverAddress) throws IOException {
+    TTransport tTransport = AuthenticationFactory.createTSocket(serverAddress);
+    switch (mAuthType) {
+      case NOSASL:
+        return new TFramedTransport(tTransport);
+      case SIMPLE:
```
Comment (apc999) [inline common/src/main/java/tachyon/security/authentication/AuthenticationFactory.java]: If indented, please comment `// indent to fall through` after `case SIMPLE`

File: common/src/main/java/tachyon/security/PlainSaslHelper.java [hunk 13]
```diff
@@ -67,7 +74,46 @@ public static TTransportFactory getPlainServerTransportFactory(AuthType authType
...
+        new PlainClientCallbackHandler(username, password), wrappedTransport);
+  }
+
+  public static class PlainClientCallbackHandler implements CallbackHandler {
+
+    private final String mUserName;
+    private final String mPassword;
+
+    public PlainClientCallbackHandler(String userName, String password) {
+      this.mUserName = userName;
```
Comment (apc999) [inline common/src/main/java/tachyon/security/PlainSaslHelper.java]: NIT: we don't use `this` in Tachyon code when possible

File: common/src/main/java/tachyon/security/PlainSaslHelper.java [hunk 14]
```diff
@@ -67,7 +74,46 @@ public static TTransportFactory getPlainServerTransportFactory(AuthType authType
...
+      this.mPassword = password;
+    }
+
+    @Override
+    public void handle(Callback[] callbacks) throws IOException, UnsupportedCallbackException {
+      for (Callback callback : callbacks) {
+        if (callback != null) {
```
Comment (apc999) [inline common/src/main/java/tachyon/security/PlainSaslHelper.java]: NIT: is `callback == null` a possible case? If not, we'd better `LOG.error` to help debugging, rather than continue silently.

File: common/src/test/java/tachyon/security/authentication/AuthenticationFactoryTest.java [hunk 15]
```diff
@@ -131,7 +121,9 @@ public void simpleAuthenticationTest() throws Exception {
     startServerThread(mTachyonConf);
 
     // when connecting, authentication happens. It is a no-op in Simple mode.
-    TTransport client = createClient(mTachyonConf, "anyone", "whatever");
+    TTransport client =
+        PlainSaslHelper.getPlainClientTransport("anyone", "whatever",
+            AuthenticationFactory.createTSocket(mServerAddress));
```
Comment (apc999) [inline common/src/test/java/tachyon/security/authentication/AuthenticationFactoryTest.java]: `AuthenticationFactory.createTSocket(mServerAddress)` is called multiple times to create a `TSocket`  in different test cases. Is it possible to make it a member variable, instead of creating it each time?

File: common/src/test/java/tachyon/security/PlainClientCallbackHandlerTest.java [hunk 16]
```diff
@@ -0,0 +1,118 @@
...
+    String user = "tachyon-user-2";
+    String password = "tachyon-user-2-password";
+    CallbackHandler clientCBHandler =
+        new PlainSaslHelper.PlainClientCallbackHandler(user, password);
+    clientCBHandler.handle(callbacks);
+
+    validateCallbacks(user, password, callbacks);
+  }
+
+
```
Comment (apc999) [inline common/src/test/java/tachyon/security/PlainClientCallbackHandlerTest.java]: NIT: one extra line here.

File: common/src/main/java/tachyon/security/PlainSaslHelper.java [hunk 17]
```diff
@@ -67,7 +74,46 @@ public static TTransportFactory getPlainServerTransportFactory(AuthType authType
...
+   * @return Wrapped Transport with Plain
+   * @throws SaslException if an AuthenticationProvider is not found
+   */
+  public static TTransport getPlainClientTransport(String username, String password,
+      TTransport wrappedTransport) throws SaslException {
+    return new TSaslClientTransport("PLAIN", null, null, null, new HashMap<String, String>(),
+        new PlainClientCallbackHandler(username, password), wrappedTransport);
+  }
+
+  public static class PlainClientCallbackHandler implements CallbackHandler {
```
Comment (apc999) [inline common/src/main/java/tachyon/security/PlainSaslHelper.java]: Could you javadoc this public static class? E.g., what it handles ?

File: common/src/main/java/tachyon/security/authentication/AuthenticationFactory.java [hunk 18]
```diff
@@ -137,6 +143,37 @@ public TTransportFactory getServerTransportFactory() throws SaslException {
     }
   }
 
-  // TODO: get client side TTransport based on auth type
-  // public TTransport getClientTransport(InetSocketAddress serverAddress) throws SaslException {}
+  /**
+   * Create a transport per the connection options. Supported transport options are: NOSASL,
+   * SIMPLE, CUSTOM, KERBEROS
```
Comment (apc999) [inline common/src/main/java/tachyon/security/authentication/AuthenticationFactory.java]: My understanding is, with `NOSASL` as input, an unmodified `TTransport` is returned; with `SIMPLE` as input, a `PlainCLientTransport` is returned, correct? It is better to javadoc those differences, just to be more explicit.

**Comment by apc999** [comment]
@sundapeng  Thanks Dapeng, this PR further cleans the code--really appreciate it!
It looks generally good, I left some comments mostly about style and documentation.
Thanks!

File: common/src/main/java/tachyon/security/authentication/AuthenticationFactory.java (hunk 12 above)
Comment (sundapeng) [inline common/src/main/java/tachyon/security/authentication/AuthenticationFactory.java]: Fixed

File: common/src/main/java/tachyon/security/PlainSaslHelper.java (hunk 13 above)
Comment (sundapeng) [inline common/src/main/java/tachyon/security/PlainSaslHelper.java]: Fixed

File: common/src/main/java/tachyon/security/PlainSaslHelper.java (hunk 14 above)
Comment (sundapeng) [inline common/src/main/java/tachyon/security/PlainSaslHelper.java]: Fixed, when callback is null, it will throw an unsupported callback exception,

File: common/src/test/java/tachyon/security/authentication/AuthenticationFactoryTest.java (hunk 15 above)
Comment (sundapeng) [inline common/src/test/java/tachyon/security/authentication/AuthenticationFactoryTest.java]: Fixed, good suggestion.

File: common/src/test/java/tachyon/security/PlainClientCallbackHandlerTest.java (hunk 16 above)
Comment (sundapeng) [inline common/src/test/java/tachyon/security/PlainClientCallbackHandlerTest.java]: Fixed

File: common/src/main/java/tachyon/security/PlainSaslHelper.java (hunk 17 above)
Comment (sundapeng) [inline common/src/main/java/tachyon/security/PlainSaslHelper.java]: Fixed. I add a comment:  A client side callback to put application provided username/password into SASL transport.

File: common/src/main/java/tachyon/security/authentication/AuthenticationFactory.java (hunk 18 above)
Comment (sundapeng) [inline common/src/main/java/tachyon/security/authentication/AuthenticationFactory.java]: Yes, thank you for your comments. I have added the javadoc.

**Comment by sundapeng** [comment]
@apc999 Thanks Bin for your review. All the comments are meaningful. If my update still have any problems, please feel free to let me know.

File: common/src/main/java/tachyon/security/PlainSaslHelper.java [hunk 19]
```diff
@@ -67,7 +74,47 @@ public static TTransportFactory getPlainServerTransportFactory(AuthType authType
     return saslFactory;
   }
 
-  // TODO: get client side PLAIN TTransport
-//  public static TTransport getPlainClientTransport(String username, String password,
-//      TTransport wrappedTransport) throws SaslException {}
+  /**
+   * Get a ClientTransport for Plain
```
Comment (ooq) [inline common/src/main/java/tachyon/security/PlainSaslHelper.java]: Can you complete the sentence?

File: common/src/main/java/tachyon/security/PlainSaslHelper.java [hunk 20]
```diff
@@ -67,7 +74,47 @@ public static TTransportFactory getPlainServerTransportFactory(AuthType authType
...
+   * @param username User Name of PlainClient
+   * @param password Password of PlainClient
+   * @param wrappedTransport The original Transport
+   * @return Wrapped Transport with Plain
```
Comment (ooq) [inline common/src/main/java/tachyon/security/PlainSaslHelper.java]: Transport -> transport, with Plain what?

File: common/src/test/java/tachyon/security/PlainClientCallbackHandlerTest.java [hunk 21]
```diff
@@ -0,0 +1,120 @@
...
+    callbacks[1] = new PasswordCallback("Password:", true);
+    callbacks[2] = new RealmCallback("Realm:");
+
+    String user = "tachyon-user-2";
+    String password = "tachyon-user-2-password";
+    CallbackHandler clientCBHandler =
+        new PlainSaslHelper.PlainClientCallbackHandler(user, password);
+    clientCBHandler.handle(callbacks);
+
+    validateCallbacks(user, password, callbacks);
```
Comment (ooq) [inline common/src/test/java/tachyon/security/PlainClientCallbackHandlerTest.java]: Guess we don't need this line.

File: common/src/test/java/tachyon/security/PlainClientCallbackHandlerTest.java [hunk 22]
```diff
@@ -0,0 +1,120 @@
...
+  }
+
+  @Test
+  public void nullPasswordCallbackTest() throws Exception {
```
Comment (ooq) [inline common/src/test/java/tachyon/security/PlainClientCallbackHandlerTest.java]: So we want to test password being `null` is fine?

File: common/src/main/java/tachyon/security/PlainSaslHelper.java [hunk 23]
```diff
@@ -67,7 +74,47 @@ public static TTransportFactory getPlainServerTransportFactory(AuthType authType
...
+    }
+
+    @Override
+    public void handle(Callback[] callbacks) throws IOException, UnsupportedCallbackException {
+      for (Callback callback : callbacks) {
+        if (callback instanceof NameCallback) {
+          NameCallback nameCallback = (NameCallback) callback;
+          nameCallback.setName(mUserName);
+        } else if (callback instanceof PasswordCallback) {
+          PasswordCallback passCallback = (PasswordCallback) callback;
```
Comment (ooq) [inline common/src/main/java/tachyon/security/PlainSaslHelper.java]: On the server, we actually disallows `null` name or `null` password, maybe we should disallow that here as well?

File: common/src/test/java/tachyon/security/PlainClientCallbackHandlerTest.java (hunk 22 above)
Comment (ooq) [inline common/src/test/java/tachyon/security/PlainClientCallbackHandlerTest.java]: Nevertheless, we should also test `null` for username.

File: common/src/test/java/tachyon/security/PlainClientCallbackHandlerTest.java [hunk 24]
```diff
@@ -0,0 +1,120 @@
...
+    callbacks[1] = new PasswordCallback("Password:", true);
+    callbacks[2] = null;
+
+    String user = "tachyon-user-4";
+    String password = "tachyon-user-4-password";
+    CallbackHandler clientCBHandler =
+        new PlainSaslHelper.PlainClientCallbackHandler(user, password);
+    clientCBHandler.handle(callbacks);
+
+    validateCallbacks(user, password, callbacks);
```
Comment (ooq) [inline common/src/test/java/tachyon/security/PlainClientCallbackHandlerTest.java]: Same, this line is not needed.

**Comment by ooq** [comment]
@sundapeng Thanks for the update, it looks quite clean now. I have some additional comments and questions.

File: common/src/test/java/tachyon/security/PlainClientCallbackHandlerTest.java (hunk 24 above)
Comment (sundapeng) [inline common/src/test/java/tachyon/security/PlainClientCallbackHandlerTest.java]: Fixed

File: common/src/main/java/tachyon/security/PlainSaslHelper.java (hunk 19 above)
Comment (sundapeng) [inline common/src/main/java/tachyon/security/PlainSaslHelper.java]: Fixed

File: common/src/main/java/tachyon/security/PlainSaslHelper.java (hunk 20 above)
Comment (sundapeng) [inline common/src/main/java/tachyon/security/PlainSaslHelper.java]: Fixed

File: common/src/test/java/tachyon/security/PlainClientCallbackHandlerTest.java (hunk 21 above)
Comment (sundapeng) [inline common/src/test/java/tachyon/security/PlainClientCallbackHandlerTest.java]: Fixed

File: common/src/test/java/tachyon/security/PlainClientCallbackHandlerTest.java (hunk 22 above)
Comment (sundapeng) [inline common/src/test/java/tachyon/security/PlainClientCallbackHandlerTest.java]: Okay, I will add a test for null username.

File: common/src/main/java/tachyon/security/PlainSaslHelper.java (hunk 23 above)
Comment (sundapeng) [inline common/src/main/java/tachyon/security/PlainSaslHelper.java]: Hi Qifan, we have disallowed null name or null password at server side, client side may not need to add this check.

**Comment by sundapeng** [comment]
Thank @ooq for your review. Very appreciate your and Bin's help.

**Comment by ooq** [comment]
LGTM. Thanks for the work! @sundapeng

**Comment by apc999** [comment]
+1 LGTM

**Comment by sundapeng** [comment]
Thank @apc999 and @ooq again for your review.
